plugins/ksp/src/org/jetbrains/kotlin/ksp/symbol/impl/binary/KSAnnotationDescriptorImpl.kt

```diff
@@ -34,27 +37,37 @@ class KSAnnotationDescriptorImpl(val descriptor: AnnotationDescriptor) : KSAnnot
     override fun <D, R> accept(visitor: KSVisitor<D, R>, data: D): R {
         return visitor.visitAnnotation(this, data)
     }
-}
 
-/**
- * Type information will be lost in KSValueArgument anyway.
- */
-fun <T> ConstantValue<T>.toValue(): Any? {
-    return when (this) {
-        is ArrayValue -> value.map { it.toValue() }
-        else -> value
+    override fun toString(): String {
+        return "@${shortName.asString()}"
     }
 }
 
+private fun ClassId.findKSClassDeclaration(): KSClassDeclaration? {
+    val ksName = KSNameImpl(this.asSingleFqName().asString())
+    return ResolverImpl.instance.getClassDeclarationByName(ksName)
+}
+
+private fun ClassId.findKSType(): KSType? = findKSClassDeclaration()?.asStarProjectedType()
+
+private fun <T> ConstantValue<T>.toValue(): Any? = when (this) {
+    is AnnotationValue -> KSAnnotationDescriptorImpl.getCached(value)
+    is ArrayValue -> value.map { it.toValue() }
+    is EnumValue -> value.first.findKSClassDeclaration()?.declarations?.find {
+        it is KSEnumEntryDeclaration && it.simpleName.asString() == value.second.asString()
+    }?.let { (it as KSClassDeclaration).asStarProjectedType() }
+    is KClassValue -> when (val classValue = value) {
+        is KClassValue.Value.NormalClass -> classValue.classId
+        is KClassValue.Value.LocalClass -> (classValue.type.constructor.declarationDescriptor as? ClassDescriptor)?.classId
```

Comment:
Nit: can you add a test case for this?

Reply:
Done. Also simplified the implementation a little bit.